src/dist/replication/lib/mutation_log.cpp

```diff
@@ -52,72 +52,190 @@ using namespace ::dsn::service;
+                }
+                else
+                {
+                    _slock.unlock();
+                }
             }
+            else
+            {
+                _slock.unlock();
+                break;
+            }    
+        }
+    }
+}
+
+
+void mutation_log_shared::write_pending_mutations(bool release_lock)
+{
+    dassert(release_lock, "lock must be hold at this point");
+    dassert(_pending_write != nullptr, "");
+    dassert(_issued_write.expired(), "");
+
+    _issued_write = _pending_write;
+    auto pr = mark_new_offset(_pending_write->size(), false);
+    dassert(pr.second == _pending_write_start_offset, "");
+
+    auto pwu = std::move(_pending_write_callbacks);
+    _pending_write_callbacks = nullptr;
+
+    auto pmu = std::move(_pending_write_mutations);
+    _pending_write_mutations = nullptr;
+
+    auto blk = std::move(_pending_write);
+    _pending_write = nullptr;
+
+    auto soffset = _pending_write_start_offset;
+
+    // seperate commit_log_block from within the lock
+    _slock.unlock();
 
-            auto hdr = (log_block_header*)(blk->data()[0].data());
+    _issued_write_task = pr.first->commit_log_block(
+        *blk,
+        soffset,
+        LPC_WRITE_REPLICATION_LOG_SHARED,
+        this,
+        [this,
+        lf = pr.first,
+        block = blk,
+        callbacks = std::move(pwu),
+        mus = std::move(pmu)
+        ](error_code err, size_t sz) mutable
+        {
+            auto hdr = (log_block_header*)block->front().data();
             dassert(hdr->magic == 0xdeadbeef, "header magic is changed: 0x%x", hdr->magic);
 
-            dinfo("end append shared log for offset = %" PRIu64, offset);
-            delete blk;
+            if (err == ERR_OK)
+            {
+                dassert(sz == block->size(),
+                    "log write size must equal to the given size: %d vs %d",
+                    (int)sz,
+                    block->size()
+                );
+
+                dassert(hdr->length + sizeof(log_block_header) == sz, "");
+
+                // flush to ensure that there is no gap between share log and in-memory buffer
+                // so that we can get all mutations in learning process.
+                //
+                // FIXME : the file could have been closed
+                lf->flush();
+            }
+
+            // notify the callbacks
+            for (auto& c : *callbacks)
+            {
+                c->enqueue_aio(err, sz);
+            }
+
+            // start to write if possible
+            if (err == ERR_OK)
+            {
+                _slock.lock();
+
+                block = nullptr;
+                if (_pending_write)
+                {
+                    write_pending_mutations(true);
+                }
+                else
+                {
+                    _slock.unlock();
+                }
+            }
         },
-        gpid_to_hash(mu->data.header.pid)
+        0
```

Comment:
ah, ```LPC_WRITE_REPLICATION_LOG_SHARED``` needs to be dispatched to the non-replication thread pool now as we now further enqueue the real callbacks in it. Oh, if we can make ```flush``` unnecessary (e.g., using direct io or something), we can ```allow_inline``` this task to reduce the latency. Let's fix the configurations together with when the flush thing is fixed.